src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -775,37 +790,124 @@ static BOOL AttributeHasChangesOmittedQuality(MTRAttributePath * attributePath)
     // 4. Cache has no entry
     // TODO: add option for BaseSubscriptionCallback to report during priming, to reduce when case 4 is hit
     if (!attributeIsSpecified || ![self _subscriptionAbleToReport] || hasChangesOmittedQuality || !attributeValueToReturn) {
+        // Read requests container will be a mutable array of items, each being an array containing:
+        //   [endpoint ID, cluster ID, attribute ID, params]
+        // Batching handler should only coalesce when params are equal.
+
+        // For this single read API there's only 1 array item. Use NSNull to stand in for nil params for easy comparison.
+        NSArray * readRequestData = @[ endpointID, clusterID, attributeID, params ?: [NSNull null] ];
+
+        // But first, check if a duplicate read request is already queued and return
+        if ([_asyncCallbackWorkQueue isDuplicateForTypeID:MTRDeviceWorkItemDuplicateReadTypeID workItemData:readRequestData]) {
+            return attributeValueToReturn;
+        }
+
+        NSMutableArray<NSArray *> * readRequests = [NSMutableArray arrayWithObject:readRequestData];
+
         // Create work item, set ready handler to perform task, then enqueue the work
         MTRAsyncCallbackQueueWorkItem * workItem = [[MTRAsyncCallbackQueueWorkItem alloc] initWithQueue:self.queue];
+        MTRAsyncCallbackBatchingHandler batchingHandler = ^(id opaqueDataCurrent, id opaqueDataNext, BOOL * fullyMerged) {
+            NSMutableArray<NSArray *> * readRequestsCurrent = opaqueDataCurrent;
+            NSMutableArray<NSArray *> * readRequestsNext = opaqueDataNext;
+
+            *fullyMerged = NO;
+
+            // Can only read up to 9 paths at a time, per spec
+            if (readRequestsCurrent.count >= 9) {
+                MTR_LOG_DEFAULT("%@ batching cannot add more", logPrefix);
+                return;
+            }
+
+            while (readRequestsNext.count) {
+                // if params don't match then they cannot be merged
+                if (![readRequestsNext[0][MTRDeviceReadRequestFieldParamsIndex]
+                        isEqual:readRequestsCurrent[0][MTRDeviceReadRequestFieldParamsIndex]]) {
+                    MTR_LOG_DEFAULT("%@ batching merged all possible items", logPrefix);
+                    return;
+                }
+
+                // merge the next item's first request into the current item's list
+                [readRequestsCurrent addObject:readRequestsNext[0]];
```

Comment:
Hmm, I made the two separate so that each is a smaller bit of logic to maintain and is clear in purpose. It also should have the added benefit of not queuing duplicates to de-dup later.

Also, if we're just checking for duplicates at batching time then we could still have duplicate read requests beyond the first 9 paths we look at for batching.

Reply:
That last is a _very_ good point.  OK, good.

In theory there's issues around duplicates being dropped and then the thing that they were duplicates of being canceled, but in practice we don't cancel things on a one-by-one basis anyway.

Reply:
I see what you mean. In the case one-by-one canceling becomes supported we can revisit the de-dup strategy then.